### PR TITLE
Add identify CJK character for word count

### PR DESCRIPTION
**Pull request by hiihellox10**
###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6006)

### CODE REVIEW

File: src/docfx/lib/HtmlUtility.cs [hunk 1]
```diff
@@ -395,34 +402,52 @@ private static bool IsLink(ref HtmlToken token, in HtmlAttribute attribute)
...
-            if (word)
+            if (isEnglishWord)
             {
                 total++;
             }
 
             return total;
         }
+
+        private static bool IsCJKChar(char c)
```
Comment (yufeih) [inline src/docfx/lib/HtmlUtility.cs:448]: This part of code is very performance sensitive (we are processing a huge number of markdown files like <https://github.com/MicrosoftDocs/azure-docs> and every character goes through this function), prefer `<` and `>` over regex.

File: src/docfx/lib/HtmlUtility.cs [hunk 2]
```diff
@@ -395,34 +402,52 @@ private static bool IsLink(ref HtmlToken token, in HtmlAttribute attribute)
...
-            var word = false;
+            var isEnglishWord = false;
 
             foreach (var ch in text)
             {
-                if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r')
+                if (IsCJKChar(ch))
                 {
-                    if (word)
+                    total++;
```
Comment (yufeih) [inline src/docfx/lib/HtmlUtility.cs:411]: Add some tests in `HtmlUtilityTest.CountWord` to cover this part.

File: src/docfx/lib/HtmlUtility.cs [hunk 3]
```diff
@@ -395,34 +402,52 @@ private static bool IsLink(ref HtmlToken token, in HtmlAttribute attribute)
         private static int CountWordInText(ReadOnlySpan<char> text)
         {
             var total = 0;
-            var word = false;
+            var isEnglishWord = false;
```
Comment (yufeih) [inline src/docfx/lib/HtmlUtility.cs:405]: Technically this is not limited to `English`. All non CJK words fall under this category, so keep the existing name?

```suggestion
            var word = false;
```

**Comment by yufeih** [comment]
@hiihellox10 Thank you for working on this. Just curious, are you working on this as a community contributor or is it part of our planed work [here](https://ceapex.visualstudio.com/Engineering/_workitems/edit/139732). I'm asking because you are part of MSFT org (but I cannot tie GitHub account back to MSFT alias) and we just happened to start working on a couple of word count features 😊

**Comment by hiihellox10** [comment]
@yufeih I'm a vendor at Alice's team, our need doc word count metadata to improve docs search api result.

**Comment by yufeih** [comment]
I see and have checked with @killa1218. This might not be directly related to https://ceapex.visualstudio.com/Engineering/_workitems/edit/99452/ (Given that task is still in feature design), but it is never the less a great improvement!

File: src/docfx/lib/HtmlUtility.cs (hunk 1 above)
Comment (hiihellox10) [reply src/docfx/lib/HtmlUtility.cs:448]: I have no better way than this, CJK character is so many, regex is the most commonly used.

File: src/docfx/lib/HtmlUtility.cs [hunk 4]
```diff
@@ -94,6 +95,13 @@ internal static class HtmlUtility
             { "iframe", new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "allowfullscreen", "height", "src", "width" } },
         };
 
+        private static readonly List<string> s_cjkCharUnicodeRangeList = new List<string>
+        {
+            "^[\u2E80-\u9FFF]$",    // CJK character
```
Comment (yufeih) [inline src/docfx/lib/HtmlUtility.cs:100]: This regex basically means any character whose char code is between `\u2E80` and `\u9FFF`, so this is equivalent to `c >= '\u2E80' && c <= '\u9FFF'`

File: src/docfx/lib/HtmlUtility.cs [hunk 5]
```diff
@@ -402,42 +404,38 @@ private static bool IsLink(ref HtmlToken token, in HtmlAttribute attribute)
...
                     {
-                        isEnglishWord = false;
+                        word = false;
                         total++;
                     }
                 }
                 else
                 {
-                    if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r')
+                    if (s_wordSeparatorChars.Any(i => i == ch))
```
Comment (yufeih) [inline src/docfx/lib/HtmlUtility.cs:423]: Revert this change because `Enumerable.Any` allocates memory and is much slower than the old version. Same as line 431.

Please also use the same technique to implement `IsCJKChar` by inlining the range check directly in line 411
